NeoMathEngine/src/CPU/CpuMathEngineDnnDropout.cpp

```diff
@@ -17,19 +17,51 @@ limitations under the License.
 #pragma hdrstop
 
 #include <MathEngineDnnDropout.h>
+#include <MemoryHandleInternal.h>
 #include <CpuMathEngine.h>
+#include <CpuMathEnginePrivate.h>
 #include <CpuExecutionScope.h>
 #include <CpuRandom.h>
 
 namespace NeoML {
 
-void CCpuMathEngine::Dropout( const CDropoutDesc& dropoutDesc, const CFloatHandle& inputData, const CFloatHandle& outputData )
+CDropoutDesc* CCpuMathEngine::InitDropout(float rate, bool isSpatial, bool isBatchwise)
+{
+	ASSERT_EXPR(rate >= 0.f && rate < 1.f);
+	auto seedDesc = new CSeedDropoutDesc(mathEngine(), true);
+	seedDesc->ForwardRate = 1.f - rate;
+	seedDesc->IsSpatial = isSpatial;
+	seedDesc->IsBatchwise = isBatchwise;
+	seedDesc->IsValid = false;
+	seedDesc->Value = 1.f / seedDesc->ForwardRate;
+	seedDesc->Threshold = (unsigned int)(seedDesc->ForwardRate * UINT_MAX);
+	return seedDesc;
+}
+
+void CCpuMathEngine::UpdateDropout(CDropoutDesc* dropoutDesc, const CBlobDesc* input,
+	const CBlobDesc* output, int seed, bool valid)
+{
+	ASSERT_EXPR(dropoutDesc != nullptr);
```

Comment:
Maybe method `Update` can be created in classes `CSeedDropoutDesc` and `CMaskDropoutDesc` to avoid  code coping